src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersisterProvider.cs

```diff
@@ -38,14 +39,9 @@ internal sealed class LanguageSettingsPersisterProvider : IOptionPersisterProvid
         public async ValueTask<IOptionPersister> GetOrCreatePersisterAsync(CancellationToken cancellationToken)
         {
             if (_lazyPersister is not null)
-            {
                 return _lazyPersister;
-            }
 
-            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-
-            var textManager = (IVsTextManager4?)await _serviceProvider.GetServiceAsync(typeof(SVsTextManager)).ConfigureAwait(true);
-            Assumes.Present(textManager);
+            var textManager = await _serviceProvider.GetServiceAsync<SVsTextManager, IVsTextManager4>().ConfigureAwait(false);
```

Comment:
❗ 

> `GetServiceAsync<SVsTextManager, IVsTextManager4>`

This extension method cannot work from unit tests. Are we absolutely sure we never need this path in unit tests?

Reply:
can you clarify that?  how did the original code work, but not necessarily the extension? thanks!

Reply:
➡️ Verified [in source](https://dev.azure.com/devdiv/DevDiv/_git/VS?path=%2Fsrc%2Fenv%2Fshell%2FPackageFramework%2FCurrent%2FShell%2FServiceExtensions.cs&version=GBmain&line=50&lineEnd=51&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) that these extensions access `ThreadHelper.JoinableTaskFactory`. This method can't be accessed in a sane way from tests, so at this time `ServiceProviderExtensions` is not usable.

Reply:
@AArnott We'd like to use ServiceProviderExtensions to clean up our code, but it uses the wrong JTF instance currently. Is there a planned API change we should be watching for or is it best to just mark ServiceProviderExtensions as a banned API?

Reply:
> how did the original code work

The original code used our JTF instance. The extension method uses a singleton from a different location.

Reply:
:'(

Reply:
What if I provided [a way](https://aka.ms/vssdktestfx) for your tests to permit access to `ThreadHelper.JoinableTaskFactory`?

Reply:
@AArnott in one way or another, that strategy violates our test isolation requirements. Examples include:

1. Every test its own specially-crafted JTF instance. Tests that are not declared as needing a UI thread validate that no calls to `SwitchToMainThreadAsync` occurred, and tests that need a UI thread are allowed to use one created for the test. `ThreadHelper` would need to use the correct instance for all tests.
2. Tests are not allowed to start asynchronous fire-and-forget work which continues running after the end of the test. We invalidate the JTF instance to ensure tests cannot cross this boundary. `ThreadHelper` allows a test to silently switch from one instance of JTF to a different instance of JTF by accessing the property at different points in time. For a static property scenario, the SDK would be required to create a new AppDomain or ALC for each test to even have a chance of meeting requirements.

Reply:
I respect those reasons. Would you prefer to see GetServiceAsync overloads added that accept `JoinableTaskFactory` instances from the caller?

Reply:
If it's not possible/expected to obtain a JTF instance from the current context automatically (allowing the caller to implicitly provide one), a suitable alternative would be overloads that accept it as a parameter.